homeassistant/components/evohome.py

```diff
@@ -0,0 +1,667 @@
+# - "400 Client Error: Bad Request for url" (e.g. Bad credentials)
+# - "429 Client Error: Too Many Requests for url" (api usuage limit exceeded)
+# - "503 Client Error: Service Unavailable for url" (e.g. website down)
+        elif err_type == "HTTPError":
+            if str(HTTP_TOO_MANY_REQUESTS) in str(err):
+                # v2 api limit has been exceeded
+                old_scan_interval = domain_data['params'][CONF_SCAN_INTERVAL]
+                new_scan_interval = max(old_scan_interval * 2, 300)
+                domain_data['params'][CONF_SCAN_INTERVAL] = new_scan_interval
+
+                _LOGGER.warning(
+                    "v2 API rate limit has been exceeded, suspending polling "
+                    "for %s seconds, & increasing '%s' from %s to %s seconds.",
+                    new_scan_interval * 3,
+                    CONF_SCAN_INTERVAL,
+                    old_scan_interval,
+                    new_scan_interval
+                )
+
+                domain_data['timers']['statusUpdated'] = datetime.now() + \
+                    timedelta(seconds=new_scan_interval * 3)
+
+            elif str(HTTP_SERVICE_UNAVAILABLE) in str(err):
+                # this appears to be common with Honeywell servers
+                pass
+
+    @callback
+    def _connect(self, packet):
+        """Process a dispatcher connect."""
+        if packet['to'] & self._type:
+            if packet['signal'] == 'update':
+                self.async_schedule_update_ha_state(force_refresh=True)
+
+    @property
+    def name(self):
+        """Return the name to use in the frontend UI."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend UI."""
+        return self._icon
+
+    @property
+    def available(self):
+        """Return True is the device is available.
+
+        All evohome entities are initially unavailable. Once HA has started,
+        state data is then retrieved by the Controller, and then the slaves
+        will get a state (e.g. operating_mode, current_temperature).
+
+        However, evohome entities can become unavailable for other reasons.
+        """
+        no_recent_updates = self._timers['statusUpdated'] < datetime.now() - \
+            timedelta(seconds=self._params[CONF_SCAN_INTERVAL] * 3.1)
+
+        if no_recent_updates:
+            # unavailable because no successful update()s (but why?)
+            self._available = False
+            debug_code = '0x01'
+
+        elif not self._status:  # self._status == {}
+            # unavailable because no status (but how? other than at startup?)
+            self._available = False
+            debug_code = '0x02'
+
+        else:  # is available
+            self._available = True
+
+        if not self._available and \
+                self._timers['statusUpdated'] != datetime.min:
+            # this isn't the first (un)available (i.e. after STARTUP), so...
+            _LOGGER.warning(
```

Comment:
No logging in properties

Reply:
Can I ask why not? This is a warning, which I feel should be logged. I know that he can tell you whether it devices available or not, but only this morning will tell you _why_ it was unavailable.

Is is it because it is a warning, rather than a debug. Or is there a general rule about doing any logging in properties, as opposed to methods.

Reply:
It's a rule here. I'm still new-ish to the project so I can't give you the ins and outs of it, but really properties should just be returning a value. There shouldn't be any reason to log anything in a property since the value they will return should be determined during the update and logging can be done there if necessary

Reply:
@jeradM Currently, I am not willing to remove this `_LOGGER.warning()` - how should we resolve the issue?

Reply:
It now looks like this:
```python
    @property
    def available(self):
        """Return True is the device is available.

        All evohome entities are initially unavailable. Once HA has started,
        state data is then retrieved by the Controller, and then the slaves
        will get a state (e.g. operating_mode, current_temperature).

        However, evohome entities can become unavailable for other reasons.
        """
        return self._available
```

Reply:
OK, now logs a message only if was previously `self._available = True`.